### ISSUE TITLE
`from_iter_instead_of_collect` feels like a blanket ban of the `FromIterator` trait

### ISSUE THREAD

**Issue by Nemo157** [rust-lang/rust-clippy#7213]
I tried this code:

```rust
Vec::from_iter(2..5)
```

I expected to see this happen: no warning

Instead, this happened:

```
warning: usage of `FromIterator::from_iter`
   --> src/lib.rs:219:1
    |
219 | Vec::from_iter(2..5);
    | ^^^^^^^^^^^^^^^^^^^^ help: use `.collect()` instead of `::from_iter()`: `(2..5).collect::<Vec<_>>()`
    |
    = note: `#[warn(clippy::from_iter_instead_of_collect)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#from_iter_instead_of_collect
```

The `Vec::from_iter` call is much more succinct and readable than `.collect::<Vec<_>>()`, and <https://github.com/rust-lang/rfcs/pull/3114> is currently in FCP-disposition-merge to add `std::iter::FromIterator` to the Rust 2021 prelude. Having Clippy recommend against using it seems to go against the decision there that it is a trait to be used directly.

I'm really not sure if there's some subset of this lint that would feel ok, there's maybe some sort of "`collect` here would be more readable", but defining heuristics to detect when to enable that seems complicated.

### Meta

- `cargo clippy -V`: `clippy 0.1.53 (42816d6 2021-04-24)`

**Comment by djc** [comment]
Previously: https://github.com/rust-lang/rust-clippy/issues/6550. Let's say it's not a coincidence that I got involved in the prelude changes. Given that the RFC has been accepted, I do think it's time to make this allow by default.

**Comment by camsteffen** [comment]
Congrats @djc on your success with appealing your case to the Supreme Court! 😉 

I agree. This lint is founded on a style principle that is, well, going out of style.

**Comment by camsteffen** [comment]
I think it would be better to fix the lint to only include cases where turbofish is not needed. Moving to pedantic doesn't really solve the issue here.

**Comment by djc** [comment]
What semantics do you suggest exactly? The lint should only trigger if the full destination type can be inferred?

**Comment by camsteffen** [comment]
Right.
